url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/Statement.java

```diff
@@ -0,0 +1,610 @@
+        boolean isGet = BeansUtils.GET.equals(methodName); //$NON-NLS-1$
+        boolean isSet = BeansUtils.SET.equals(methodName); //$NON-NLS-1$
+        if (!isGet && !isSet) {
+            throw new NoSuchMethodException(Messages.getString("custom.beans.3C")); //$NON-NLS-1$
+        } else if (args.length > 0 && args[0].getClass() != Integer.class) {
+            throw new ClassCastException(Messages.getString("custom.beans.3D")); //$NON-NLS-1$
+        } else if (isGet && args.length != 1) {
+            throw new ArrayIndexOutOfBoundsException(
+                    Messages.getString("custom.beans.3E")); //$NON-NLS-1$
+        } else if (isSet && args.length != 2) {
+            throw new ArrayIndexOutOfBoundsException(
+                    Messages.getString("custom.beans.3F")); //$NON-NLS-1$
+        }
+
+        Class<?>[] paraTypes = isGet ? new Class<?>[] { Object.class, int.class }
+                : new Class<?>[] { Object.class, int.class, Object.class };
+        return Array.class.getMethod(methodName, paraTypes);
+    }
+
+    private Constructor<?> findConstructor(Class<?> clazz, Object[] args)
+            throws NoSuchMethodException {
+        Class<?>[] argTypes = getTypes(args), paraTypes, resultParaTypes;
+        Constructor<?> result = null;
+        boolean isAssignable;
+        for (Constructor<?> constructor : clazz.getConstructors()) {
+            paraTypes = constructor.getParameterTypes();
+            if (match(argTypes, paraTypes)) {
+                if (result == null) {
+                    // first time, set constructor
+                    result = constructor;
+                    continue;
+                }
+                // find out more suitable constructor
+                resultParaTypes = result.getParameterTypes();
+                isAssignable = true;
+                for (int index = 0; index < paraTypes.length; index++) {
+                    if (argTypes[index] != null
+                            && !(isAssignable &= resultParaTypes[index]
+                                    .isAssignableFrom(paraTypes[index]))) {
+                        break;
+                    }
+                    if (argTypes[index] == null
+                            && !(isAssignable &= paraTypes[index]
+                                    .isAssignableFrom(resultParaTypes[index]))) {
+                        break;
+                    }
+                }
+                if (isAssignable) {
+                    result = constructor;
+                }
+            }
+        }
+        if (result == null) {
+            throw new NoSuchMethodException(Messages.getString(
+                    "custom.beans.40", clazz.getName())); //$NON-NLS-1$
+        }
+        return result;
+    }
+
+    /**
+     * Searches for best matching method for given name and argument types.
+     */
+    static Method findMethod(Class<?> clazz, String methodName, Object[] args,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 346-354 into a separate method.